src/sonic-eventd/tools/events_monit_test.py

```diff
@@ -0,0 +1,226 @@
+#! /usr/bin/env python3
+
+from inspect import getframeinfo, stack
+from swsscommon.swsscommon import events_init_publisher, event_publish, FieldValueMap
+from swsscommon.swsscommon import event_receive_op_t, event_receive, events_init_subscriber
+from swsscommon.swsscommon import events_deinit_subscriber, events_deinit_publisher
+import argparse
+import os
+import threading
+import time
+import syslog
+import uuid
+
+chk_log_level = syslog.LOG_ERR
+
+test_source = "sonic-host"
+test_event_tag = "device-test-event"
+test_event_key = "{}:{}".format(test_source, test_event_tag)
+test_event_params = {
+    "sender": os.path.basename(__file__),
+    "reason": "monit periodic test",
+    "batch-id": str(uuid.uuid1()),
+    "index": "0"
+}
+
+# Async connection wait time in seconds.
+ASYNC_CONN_WAIT = 0.3
+RECEIVE_TIMEOUT = 1000
+
+# Thread results
+rc_test_receive = -1
+publish_cnt = 0
+
+def _log_msg(lvl, pfx, msg):
+    if lvl <= chk_log_level:
+        caller = getframeinfo(stack()[2][0])
+        fmsg = "{}:{}:{}".format(caller.function, caller.lineno, msg)
+        print("{}: {}".format(pfx, fmsg))
+        syslog.syslog(lvl, fmsg)
+
+def log_err(m):
+    _log_msg(syslog.LOG_ERR, "Err", m)
+
+def log_info(m):
+    _log_msg(syslog.LOG_INFO, "Info",  m)
+
+def log_notice(m):
+    _log_msg(syslog.LOG_NOTICE, "Notice", m)
+
+def log_debug(m):
+    _log_msg(syslog.LOG_DEBUG, "Debug", m)
+
+
+def map_dict_fvm(s, d):
+    for k, v in s.items():
+        d[k] = v
+
+
+# Invoked in a separate thread
+def test_receiver(event_obj, cnt):
+    global rc_test_receive
+
+    sh = events_init_subscriber(False, RECEIVE_TIMEOUT, None)
+
+    # Sleep ASYNC_CONN_WAIT to ensure async connectivity is complete.
+    time.sleep(ASYNC_CONN_WAIT)
+
+    exp_params = dict(test_event_params)
+
+    # Signal main thread that subscriber is ready to receive
+    event_obj.set()
+    cnt_done = 0
+
+    while cnt_done < cnt:
+        p = event_receive_op_t()
+        rc = event_receive(sh, p)
+        
+        if rc > 0 and publish_cnt < 2:
```

Comment:
>publish_cnt [](http://example.com/codeflow?start=22&length=11)

If `publish_cnt < 2`, this is infinite loop?

Reply:
On seperate thread, we have an event publisher that publishes event which will raise publish_cnt (line 154)

Reply:
This is in design. In reality, if some threads crash or not functional well, will this thread burn CPU?

Reply:
Main thread is timed. After 5 async publish calls, it waits a max of 1  second and exit.